hal/halx86/generic/kdpci.c

```diff
@@ -0,0 +1,516 @@
+
+                    case PCI_CARDBUS_BRIDGE_TYPE:
+                    {
+                        /* FIXME: Implement PCI resource allocator */
+                        break;
+                    }
+
+                    default:
+                        break;
+                }
+
+                if (!PCI_MULTIFUNCTION_DEVICE(&PciConfig))
+                    break;
+            }
+        }
+    }
+
+    return FALSE;
+}
+
+CODE_SEG("INIT")
+VOID
+NTAPI
+HalpRegisterPciDebuggingDeviceInfo(VOID)
+{
+    ULONG i;
+    NTSTATUS Status;
+    WCHAR StringBuffer[16];
+    BOOLEAN HasDebuggingDevice = FALSE;
+    UNICODE_STRING KeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\"
+                                                 L"CurrentControlSet\\Services\\PCI\\Debug");
+    HANDLE Handle, KeyHandle;
+
+    PAGED_CODE();
+
+    for (i = 0; i < RTL_NUMBER_OF(HalpPciDebuggingDevice); ++i)
+    {
+        if (HalpPciDebuggingDevice[i].u.bits.InUse)
+        {
+            HasDebuggingDevice = TRUE;
+            break;
+        }
+    }
+    if (!HasDebuggingDevice)
+    {
+        /* Nothing to register */
+        return;
+    }
+
+    Status = HalpOpenRegistryKey(&Handle, 0, &KeyName, KEY_ALL_ACCESS, TRUE);
+    if (!NT_SUCCESS(Status))
+        return;
+
+    for (i = 0; i < RTL_NUMBER_OF(HalpPciDebuggingDevice); ++i)
+    {
+        ULONG Value;
+        PCI_SLOT_NUMBER PciSlot;
+
+        if (!HalpPciDebuggingDevice[i].u.bits.InUse)
+            continue;
+
+        RtlInitEmptyUnicodeString(&KeyName, StringBuffer, sizeof(StringBuffer));
+        RtlIntegerToUnicodeString(i, 10, &KeyName);
+        Status = HalpOpenRegistryKey(&KeyHandle,
+                                     Handle,
+                                     &KeyName,
+                                     KEY_ALL_ACCESS,
+                                     TRUE);
+        if (!NT_SUCCESS(Status))
+            continue;
+
+        Value = HalpPciDebuggingDevice[i].u.bits.BusNumber;
+        RtlInitUnicodeString(&KeyName, L"Bus");
+        ZwSetValueKey(KeyHandle,
+                      &KeyName,
+                      0,
+                      REG_DWORD,
+                      &Value,
+                      sizeof(Value));
+
+        PciSlot.u.AsULONG = 0;
+        PciSlot.u.bits.DeviceNumber = HalpPciDebuggingDevice[i].u.bits.DeviceNumber;
+        PciSlot.u.bits.FunctionNumber = HalpPciDebuggingDevice[i].u.bits.FunctionNumber;
+        Value = PciSlot.u.AsULONG;
+        RtlInitUnicodeString(&KeyName, L"Slot");
+        ZwSetValueKey(KeyHandle,
+                      &KeyName,
+                      0,
+                      REG_DWORD,
+                      &Value,
+                      sizeof(Value));
+
+        ZwClose(KeyHandle);
+    }
+
+    ZwClose(Handle);
+}
+
+CODE_SEG("INIT")
+NTSTATUS
+NTAPI
+HalpReleasePciDeviceForDebugging(
+    _Inout_ PDEBUG_DEVICE_DESCRIPTOR PciDevice)
+{
+    ULONG i;
+
+    DPRINT0("%s(%p) called\n", __FUNCTION__, PciDevice);
```

Comment:
dprint 0 ?

Reply:
Early phase debugging

Reply:
Ah, with vgal's patch?

Reply:
Vgal's one looks a bit different: #3393 (it's defined as `DbgKdPrint0`)